src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs

```diff
@@ -268,23 +268,23 @@ private List<ExtractedResult> ExtractResultsFromRuns(IEnumerable<Run> sarifRuns)
                 indexRemappingVisitor.HistoricalLogicalLocations = resultPair.Run.LogicalLocations;
                 indexRemappingVisitor.VisitResult(result);
 
-                if (result.RuleIndex != -1)
+                string ruleId = result.ResolvedRuleId(resultPair.Run);
+                if (!string.IsNullOrEmpty(ruleId))
                 {
-                    ReportingDescriptor rule = result.GetRule(resultPair.Run);
-                    if (rule != null)
+                    if (reportingDescriptors.TryGetValue(ruleId, out int ruleIndex))
                     {
-                        if (reportingDescriptors.TryGetValue(rule, out int ruleIndex))
-                        {
-                            result.RuleIndex = ruleIndex;
-                        }
-                        else
-                        {
-                            result.RuleIndex = reportingDescriptors.Count;
-                            reportingDescriptors[rule] = reportingDescriptors.Count;
+                        result.RuleIndex = ruleIndex;
+                        result.RuleId = ruleId;
+                    }
+                    else
+                    {
+                        ReportingDescriptor rule = result.GetRule(resultPair.Run);
 
-                            run.Tool.Driver.Rules = run.Tool.Driver.Rules ?? new List<ReportingDescriptor>();
-                            run.Tool.Driver.Rules.Add(rule);
-                        }
+                        result.RuleIndex = reportingDescriptors.Count;
+                        reportingDescriptors[ruleId] = reportingDescriptors.Count;
+
+                        run.Tool.Driver.Rules = run.Tool.Driver.Rules ?? new List<ReportingDescriptor>();
```

Comment:
>= [](start = 46, length = 1)

Are we building SDK with C# 8 now? If so can change this to, `run.Tool.Driver.Rules ??= new List<ReportingDescriptor>();`

Reply:
Changed.